crates/sui-core/src/authority/authority_store.rs

```diff
@@ -595,6 +595,47 @@ impl AuthorityStore {
         Ok(ret)
     }
 
+    /// Load a list of objects from the store by object reference.
+    /// If they are exist in the store, they are returned directly.
+    /// If any object missing, we try to figure out the best error to return.
+    /// If the object we are asking is currently locked at a future version, we know this
+    /// transaction is out-of-dated and we return a ObjectVersionUnavailableForConsumption,
+    /// which indicates this is not retriable.
+    /// Otherwise, we return a ObjectNotFound error, which indicates this is retriable.
+    pub fn multi_get_object_with_more_accurate_error_return(
+        &self,
+        object_refs: &[ObjectRef],
+    ) -> Result<Vec<Object>, SuiError> {
+        let objects = self.multi_get_object_by_key(
+            &object_refs.iter().map(ObjectKey::from).collect::<Vec<_>>(),
+        )?;
+        let mut result = Vec::new();
+        for (object_opt, object_ref) in objects.into_iter().zip(object_refs) {
+            match object_opt {
+                None => {
+                    let lock = self.get_latest_lock_for_object_id(object_ref.0)?;
+                    let error = if lock.1 >= object_ref.1 {
```

Comment:
whats the reason why we wouldn't return `ObjectVersionUnavailableForConsumption` if the object is locked at a previous version?

Reply:
My understanding is that we want ObjectNotFound to be a retriable error because it's possible a validator is behind, and hence not observing a newer version yet.
So we want to be able to distinguish the two cases.
Although to be honest all of these are too complicated and I wonder if we shouldn't try that hard to make ObjectNotFound a retriable error in the first place.